### DISCUSSION TITLE
How to run code when any scenario fails?

### DISCUSSION THREAD

**Discussion by ms-ati** [cucumber-rs/cucumber#245] [Q&A] [answered]
Hi! I am trying to use `::after` hook or find other means, to cause some code to run when a `Scenario` fails.

In particular -- I'd like, optionally, to have a per-Scenario temporary directory **not be deleted** when the test fails, so that it is possible for the user to examine the contents of the temp dir.

**Comment by tyranron** [comment, marked as answer]
@ms-ati you have two options:
1. Use [`.after()` hook](https://cucumber-rs.github.io/cucumber/current/writing/hooks.html#after-hook). It runs even if the scenario has failed, however is enabled for all executed scenarios, which may not be convenient.
2. Use custom `Drop` implementations. This can be more tedious in terms of writing code, but allows any desired level of granularity.

> In particular -- I'd like, optionally, to have a per-Scenario temporary directory

`World` has a lifespan of a scenario. So you can just put it into your `World` as `Option<TempDir>`.

> **not be deleted** when the test fails, so that it is possible for the user to examine the contents of the temp dir.

So you should set some sort of flag at the last step of the scenario in your `World`. And if it's not set, then [`ManuallyDrop` it](https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html) in `Drop`.

**Reply by tyranron** [reply]
@ms-ati 

>  In either place, how can I determine whether the current Scenario is in the Failed state? I haven’t been able to find a channel where that information (which I see in `events.rs`) is made available either to the World Drop implementation, the after hook, or anywhere else?

This is left on the implementor choice, at the moment. If you need this - you should track it in your `World`.

For, example, as I wrote before, having a flag in your `Drop`-wrapper or `World` which you will set in the last step, and if it's not set, then the last step hasn't executed successfully, so the scenario was either failed or skipped:
>  So you should set some sort of flag at the last step of the scenario in your `World`. And if it's not set, then [`ManuallyDrop` it](https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html) in `Drop`.

----------------------

@ilslv maybe it's worth to pass [`event::Scenario`](https://docs.rs/cucumber/0.16/cucumber/event/enum.Scenario.html) to `.after()` hook?

**Reply by ms-ati** [reply]
Sorry, can you clarify @tyranron:

> For, example, as I wrote before, having a flag in your Drop-wrapper or World which you will set in the last step, and if it's not set, then the last step hasn't executed successfully, so the scenario was either failed or skipped:

1. This really does not offer a way to identify _Failed_ scenarios from Skipped, right?
2. This requires literally every `#[then]` step to set a flag or otherwise inform `World` that the Scenario as a whole succeeded?

Yeah this doesn't seem quite right to me thinking this through -- are we sure that having an `on_failure` hook, or passing the state information to a generic `after` hook, wouldn't work better?

**Reply by ilslv** [reply]
@tyranron 

> maybe it's worth to pass [`event::Scenario`](https://docs.rs/cucumber/0.16/cucumber/event/enum.Scenario.html) to .after() hook?

Yes, but I would create a separate enum for this, because [`event::Scenario::Hook`](https://docs.rs/cucumber/0.16.0/cucumber/event/enum.Scenario.html#variant.Hook) is unreachable from inside the `Hook` itself.

**Reply by tyranron** [reply]
@ilslv 

> Yes, but I would create a separate enum for this, because [`event::Scenario::Hook`](https://docs.rs/cucumber/0.16.0/cucumber/event/enum.Scenario.html#variant.Hook) is unreachable from inside the `Hook` itself.

I do think that the last [`event::Step`](https://docs.rs/cucumber/0.16.0/cucumber/event/enum.Step.html) would be enough, as we don't need anything else from [`event::Scenario`](https://docs.rs/cucumber/0.16/cucumber/event/enum.Scenario.html), actually.

------------------------------------

@ms-ati 

> 1. This really does not offer a way to identify _Failed_ scenarios from Skipped, right?

Depends on the user-side implementation. If this distinction is needed, once could always invent something, like `.catch_unwind()` wrappers, or etc.

>  2. This requires literally every `#[then]` step to set a flag or otherwise inform `World` that the Scenario as a whole succeeded?

Yup.

> Yeah this doesn't seem quite right to me thinking this through -- are we sure that having an `on_failure` hook, or passing the state information to a generic `after` hook, wouldn't work better?

It would be better, of course. It's just not implemented yet, as there was no demand for it before. The proposed ways are about how you can solve your issue here and now.

**Reply by ilslv** [reply]
@tyranron 

> I do think that the last [`event::Step`](https://docs.rs/cucumber/0.16.0/cucumber/event/enum.Step.html) would be enough, as we don't need anything else from [`event::Scenario`](https://docs.rs/cucumber/0.16/cucumber/event/enum.Scenario.html), actually.

👍

**Reply by tyranron** [reply]
@ms-ati now available in [0.17.0](https://github.com/cucumber-rs/cucumber/releases/tag/v0.17.0).

**Reply by ms-ati** [reply]
Wow... this is amazing @tyranron, looking forward to trying it soon!

**Comment by ms-ati** [comment]
Thank you @tyranron! My question to clarify is:
1. Where can we know that a Scenario failed?

Thank you for pointing out the after hook and World life cycle. In either place, how can I determine whether the current Scenario is in the Failed state? I haven’t been able to find a channel where that information (which I see in `events.rs`) is made available either to the World Drop implementation, the after hook, or anywhere else?

**Comment by ms-ati** [comment]
Hi @tyranron @ilslv, thank you so much for introducing the `ScenarioFinished` event to the `after` hook parameters!

I have it working here: https://github.com/ms-ati/aruba-rs/blob/bc225e6d14dedfaeafb36abb6144d87f99ba0e91/examples/aruba_hello_world/tests/cucumber_aruba_example.rs#L11

However, I'm still looking for a way to configure this after hook in the `aruba-rs` library of steps, rather than in the test that ultimately uses the aruba-rs library.

I cannot, however, figure out how to either call `after`, or define `cucumber`, on the `ArubaWorld` struct. Do you have any ideas?

**Reply by tyranron** [reply]
@ms-ati I think you may provide own methods/function which:
- either return already preconfigured `Cucumber` object, so the library users may only do some final tweaks;
- or do all the configurations/running without exposing `Cucumber` object to library users at all.
